### ISSUE TITLE
Mod Compatibility: Movable Intimidated Cop

### ISSUE THREAD

**Issue by Malidictus** [Kamikaze94/WolfHUD#67]
I've run into a compatibility issue between WolfHud and TdlQ's Movable Intimidated Cop mod (the one which allows you to move intimidated cops like regular hostages). When moving intimidated cop is shot in transit (which happens a lot because they can still take damage from friendly AI), the hostage counter does not go down. This is not just a cosmetic issue, either, as any such "phantom" dominated cops will count against my dominated cop maximum. During a Cook Off heist, I managed to lose four cops in transit, and as such could only dominate 4 at a time, when I should have been able to dominate 8 in a team of four people.

This happens consistently every time an intimidated cop gets shot while moving, as well as when said cop gets shot while standing still after having been moved (I presume the mod changes their unit type somehow?) when WolfHud is enabled. Disabling WolfHud returns the Movable Intimidated Cop to proper operation. Reactivating WolfHud causes the issue to happen again, but I've sometimes had it not happen on the first heist after reactivating it.

I couldn't even begin to tell you how the Movable Intimidated Cop mod works since it's way over my head. All I know is TdlQ asked me to look for a function called CopBrain:clbk_death(<parameters>) - I don't remember the exact call parameter list - which gets called when one of his movable intimidated cops gets killed and handles reducing the counter. I don't know if that helps, but it's literally the extent of my knowledge on the matter.

WolfHud had a lot of issues handling movable intimidated cops in general. A lot of times they'll be added to the hostages counter (ziptied hands icon) when first intimidated, only to be taken out of there and added to the dominated cops counter (ye olde Dominator icon) when I move them and set them down again. I don't know if that contributes to the issue, but I see it very often.

I realise this is an issue with another mod that you have no obligation to support, but I feel that the Movable Intimidated Cop mod adds in functionality that the game by all rights should have had from the start. All I can do is make this request and hope for the best.

**Comment by Tatsugi** [comment]
Offtopic, but last sentence reminds me: "Hope for the best, prepare to the worst." :D

**Comment by Malidictus** [comment]
Well, "the worst" would be Kamikadze saying a flat no :)

I did some more testing today. I don't know what causes it, but WolfHud will count dominated cops as "hostages" a lot of the time even before moving them. Just now, I attempted to dominate 4 cops. 1 counted as a hostage, 3 counted as dominated cops.

Having tested this some more, I don't think cops who are stood in place will trigger the bug, nor will cops who are set down. It's only while they're physically walking that the issue occurs. When set down, they seem to count as ordinary dominated cops. When standing still, they throw their hands up in the air and may count as cops in the process of being dominated. While walking, they use a generic "walk" animation that I don't know where it comes from. It seems like only then does the bug occur. Shooting them in that situation prevents the counter for either hostages or dominated cops (whichever they're dosplayed as) from going down.

**Comment by killrhythm09** [comment]
Just chiming in to say I've seen the behavior where it counts dominated cops as regular tied hostages. I don't use Movable Intimidated Cops, but that always bothered me a little bit.

**Comment by Kamikaze94** [comment]
The game only counts the total number of hostages (police and civilian) for all players.
The Host has the number of police hostages additionaly.
In Seven's version, HUDList would go through all enemy units and check if their surrender animation is "surrender" for clients, to divide them.
The Problem when changing police hostages no longer counting in the enemy counter was, that HUDList has 2 categories, cop and security, which both are dominatable.
So I changed that client workaround beeing used by the host too.
So everytime a hostage status changes, the game goes though the enemy list, looking for hostages, seperating them by security_hostages and cop_hostages, and saying the rest to reach the total count have to be civilian hostages.
The problem why that hostage could appear sometimes as civilian hostage is if the game is too slow, setting the animation state for the unit. There is a small delay in place, to give the base game some time, to get its stuff straight, and I rarely if ever encounter that problem, but it can at least possibly occure.

The problem with MIC is now, that it changes the cops animation state, so that mentioned function no longer detects the cop as surrendered. I changed something yesterday, to decide based on the units active logic, but that is probably Host only (+ the logic is the same as normal walking around cops.)
I even tried getting the units objective, if that one is in "follow" state. But again, I think that would only work as Host.
I havn't found an attribute yet, to properly decide, if the unit is a hostage or not, when MIC is beeing used....  (might need to check for that mods "this unit is beeing moved flag", might be again only on hosts end, not sure here. Can MIC be used as Client?)


Btw, cop hostages counting as civ hostages, when which case do you normally encounter that: 
as host or client, or both?

**Comment by killrhythm09** [comment]
Here's a perfect example of it happening literally just now, when I was hosting (using the latest dev build of WolfHud):

http://i.imgur.com/jBiO6w1.jpg

I'm pretty sure I've also seen it as a client, but I would need to test more if you'd like to see that in action as well.

**Comment by Malidictus** [comment]
I spoke with TdlQ, and he said that "unit:base().mic_is_being_moved is a good flag to use" but also that "it's not safe to use on clients," here referring to clients without MIC. He wasn't sure if clients with MIC would have that flag. I don't know what you'd do for clients without MIC at all.

And yes, the Movable Intimidated Cop mod can be used as a client, but the host would also have to be running the mod. Clients with the mod playing with a host who doesn't have it will still get an interaction prompt and can even attempt the interaction, but nothing happens. Clients without the mod playing with a host who has it can still set the moving intimidated cop down, they just can't cause him to move. The interaction still works, so presumably there's some way the game is able to tell that that's a hostage even while he's walking.

**Comment by Kamikaze94** [comment]
Yeah, have found that flag, and from my solo tries it seems to work.
Not sure how that behaves on clients. Clients without MIC will never have that flag. As the aniumation state will get changed for them too, it will probably be possible for them, that the dominated cop counts as a normal hostage. Not sure about clients that use MIC, it doesn't get syncted explicit, as far as I saw, but it might get passed over on some normal unit syncs....

On Hostage/Dominated dispaly:
Thats what I thought of. Clients have a time window in which the hostage count will be updated frequently, while on Hosts it's only doing it once, with a delay to the actual domination.
Changed that to the window variant for everyone, which will hopefully resolve the problems...
:)

**Comment by Malidictus** [comment]
I just tried the latest version of WolfHud that I got off the GitHub repo and it's still prone to bugging out even with this. It's inconsistent, as well. We surrendered one guard and he counted as a dominated cop. Killing him while he was moving didn't cause a bug. We then surrendered a blue shirt cop who counted as a hostage (ziptied wrists icon) on my side. Killing him caused a bug, as he was still counted as existing even though he was dead.

I tested this on a host running Keepers and WolfHud (that's me) and a host running WolfHud but not Keepers (that's a friend of mine). When it worked as intended, it worked for both of us. When it bugged out, it also bugged out for both of us.

---

Unrelated question, but are hostage numbers not synced up and verified in some way in the base game? As in, does the game not check every now and then to make sure the number is correct? It seems really odd to hinge hostage count on an independent counter, rather than in some way polling the actual units in the world. Or would that lag too much?

**Comment by Malidictus** [comment]
Thinking I must have done something wrong, I redownloaded WolfHud from GitHub not 15 minutes ago and retried my experiment solo. I dominated a regular guard on Jewellery Store, moved him and shot him while he was moving. The counter still counted him as alive even though he was dead. This was solo online, obviously with me as a host. I honestly don't see any change in behaviour from before the fixes.

Is it possible that I'm doing something wrong while downloading the mod such that I keep getting the wrong version? I even ran a text diff on my core.lua and the one here in GitHub and they seem to match.

**Comment by Kamikaze94** [comment]
"GameInfoManager.lua" , line 1680, should say: 
`if u_data.unit:anim_data().surrender or u_data.unit:base().mic_is_being_moved then`

At least solo it seemed to work for me.

**Comment by Malidictus** [comment]
I checked, and that is indeed what line 1680 of GameInfoManager.lua says, but it still doesn't work for me. I should say, it doesn't work every time. Sometimes I'll shoot a moving cop and the counter will decrement as expected, sometimes I'll shoot a moving cop and it won't. I tested it once on Crime.net offline, and of the two dominated I took, one bugged out when shot while moving. I tested it on Crime.net online, and the one and only dominated cop I took bugged out when shot while moving.

I think the reason I'm noticing this just now when it was apparently the case since I started using WolfHud is precisely this reason - it's inconsistent. Sometimes killing moving intimidated cops decrements the counter, sometimes it doesn't and I genuinely don't know what's causing it. This is all as a host, as well - both of the above tests and most of my experience using MIC has been as a host. Is there any way I can gather more useful information beyond just anecdotes for you?

**Comment by Kamikaze94** [comment]
Hmm, tried ever possible situation I could think of now.
(Taking one guard hostage, killing him while moving, getting the next, ... or taking multiple, moving them, taking civ hostages to update the counter killing one, taking more hostages....)
Still hasn't happened to me.

And no, I'm not even sure, what I would do to find out, why your hostages aren't in surrender animation, nor the MIC flag is set...

Not sure, if mentioned before, but at least this is just a visual bug. The police hostage count on the hosts end never gets written (or read), so its "just" the counter freaking out.
Still unsatisfying though...  :/

**Comment by Malidictus** [comment]
> Not sure, if mentioned before, but at least this is just a visual bug. The police hostage count on the hosts end never gets written (or read), so its "just" the counter freaking out.

That's definitely not a cosmetic bug, though. I ran a test about a week ago. Dominate cops one at a time, move them, shoot them. Once I'd done that to 5 of them, I became completely unable to dominate more cops whatsoever. Doesn't matter their level of health, suppression, reloading, knockback - they would not surrender. This is the behaviour I see when we've reached the cap of surrendered cops + jokers, which is 4 + the number of players. Solo, that comes up to 5, which is why I expected this behaviour after 5 of them.

Something else weird: When I surrender, move and shoot an intimidated cop, I get stuck with the counter saying "1 intimidated cop" using the handcuffs icon. If I have normal hostages on the map, though, that display will actually turn into a hostage using the ziptied hands icon after a while of trying to surrender more cops. This leads me to the suspicion that the list is in some way updated for hostages I'm not currently interacting with. In fact, I meant to ask if there's some way to either force the list to verify how many hostages we have either on its own or when interacting with them.

---

I'll try and download WolfHud from the GitHub repo again. See if that helps. I'll try to test it with no other mods of any kind and see what happens. Well, no other mods aside from Keepers, obviously :) If you feel like it, you can drop me a friend invite on Steam to test this out together. Just leaving that out there as an option, it's not necessary.

**Comment by Malidictus** [comment]
OK, I figured it out, I think. Disabling one of TdlQ's other mods - the Full Speed Swarm - made shooting moving cops work correctly. It might have been the cause all along, come to think of it. I don't know exactly how that's causing it, but my random guess would be his re-implementation of the Delayed Calls functionality. Without FSS, WolfHud counts hostages correctly now, and indeed seems to read them as the correct type (dominated cops don't count as hostages).

I've only done very limited testing (as in, one heist where I shot five moving intimidated cops), but it seems fine so far. I'll do more testing "in the wild" as time permits later today, but this is looking good!

**Comment by Malidictus** [comment]
I spoke with TdlQ. He identified and fixed an issue with his own Full-Speed Swarm mod. Along with your fix for WolfHud, movable intimidated cops now work as intended with all mods enabled. They get tagged as the correct type of hostage and they get counted correctly if killed while moving them. At least, that's solo with me as the host. I'll run more tests when my friends log in tonight, to see how the interaction works between Keepers, non-Keepers, WolfHud and non-WolfHud. So far so good.

Sorry to spam your bugs feed...

**Comment by Kamikaze94** [comment]
Yeah, thought it has to be the original GroupAiManager:state():on_hostage_state()-function not beeing called. therefore your general hostage count as well as the police hostage count freaking out.
Looked through WolfHUD, MIC and Keepers, but couldn't find where that error could possibly come from.
So, that coming from another mod makes sense.

Anyways, nice it woked out. I don't expect trouble when mixing those mod usages around. (Possibly only for non MIC users, if Host uses it, but there is no way around, unless MIC would abuse some basegame function, trying to set a flag, that is originally intended for something else, probably better not to do....^^

And it seems like I cannot get below 10 issues for long. Tried it, always when I close some, others are popping up, :P
At least they are all in one place and nicely seperated....^^

**Comment by Malidictus** [comment]
Hah, sorry to keep making new ones :) You can probably close my "crash on start" one that turned out to be alignment settings.

On topic: I tried a few games with me as the host bot solo and with other people, and the hostage counter holds steady. Bots killed A LOT of my MICs (thanks for helping me test it, I guess...), and I never at any point got a wrong number. In fact, I was able to track to fate of my MICs by the counter, knowing they've died when the number dropped. So far so good.

I didn't get the chance to test this as a client since apparently everyone is playing other stuff today, but I'll get a chance tomorrow. So far it's looking good, though. My suspicion is it's fixed.

**Comment by Malidictus** [comment]
Having run a few full heists with a friend running WolfHud but NOT Keepers, we ran into no notable bugs. Hostage counts were consistent regardless of how many dominated cops I shot while they were moving and never caused us any issue or discrepancy. I was told that dominated cops are counted as "hostages" while I'm moving them but revert to dominated cops when I set them down, which is an issue of the client not having MIC, I presume.

Basically, I've run out of ways to test this compatibility issues, and I don't see it any more. With the fix to WolfHud and FSS, everything seems to be working like it should. I'm tremendously happy with it, as I can now resume stocking up spare cops for the winter. Thank you kindly :)

**Comment by Kamikaze94** [comment]
Yay, nice it woked out.
Getting the counter to work on clients, when the host is running MIC might be tricky, because they use the default "travel" Ai state. Maybe checking against their set objective xould help, but I'd guess that those are host-side infos....
At least it works now, and you got rid of the bugs filling your domination count.

Guess I can close this now... finally, xD

**Comment by Malidictus** [comment]
Honestly, it seems like the counter updates on clients just fine. I tried it with a client who didn't have Keepers and the counter was always true, even if it sometimes mistook dominated cops for Jokers. As far as I'm concerned this is fixed. Sorry to send you messages after you closed it :)
